core/WFCORE-5343_Support_YAML_files_for_configuration_customization.adoc

```diff
@@ -0,0 +1,144 @@
+----
+Using a null value is not 'doable' as we have attrbiute where the 'null' value is significant.
+
+* Allow setting null if value is empty and attribute is null-significant.
+* Allow support to remove resource using a YAML tag !remove. For example:
+----
+wildfly-configuration:
+    subsystem:
+        microprofile-jwt-smallrye: !remove
+----
+
+* Allow support for adding element to list and supporting index: this might impact idempotence.
+
+=== Non-Requirements
+
+The goal of the YAML files is to be able to customize an existing configuration. It is not here to replace the existing configuration support with XML.
+As such we won't support part of the management model. Those elements would be filtered out:
+
+ - extension: to add extension to the server as this might require modules which can be missing.
+ - deployment: to add deployments to the server as this require more that just some configuration.
+ - deployment-overlay: to add deployment-overlays to the server as this require more that just some configuration.
+ - path: since those should already have been defined when the YAML files are parsed.
+ 
+As this configuration extension is for a *standalone server* only this *won't be supported in domain mode*.
+
+Because we need to support expressions, it migt be complex to get a JSON schema describing the YAML files for IDE completion. Also such a schema would be really big and would be really difficult to maintain or verify.
+
+== Implementation Plan
+
+If we are processing YAML files then the server is in *read-only* mode: that makes idempotence easy to acheive. We could persist the resulting configuration after boot so that we wouldn't be in read-only mode but then what is the purpose of the YAML files after the 1st execution ?
+We are going to process the YAML files to get a list of operations to be applied at boot after the operations defined in the XML configuration file. That means that when we are processing the YAML files we don't have access to the model that was persisted int the XML configuration file.
+What we are going to do is use the boot operations to define what to do with what is described in the YAML files.
+In the YAML tree we can check if the node is matching a resource or an attribute.
+If it is a resource then we check if there are boot operations for this address, if such is the case then we have to process the attributes, otherwise that means that we need to create the resource.
+To create a resource we will look into the YAML subtree for the attributes that are defined, and add an `add` operation for it.
+If the resource is created then we look for it attributes that are writeable in the YAML subtree and add a `write-attribute` operation for each of them. That means that we need to have every element in a list since we are not using `list-add` in this case.
+Then we process the sub-resources if any exist.
+We should be able to get the effective XML file to validate the resut of our YAML configuraiton.
```

Comment:
when are you planning to validate the XML? against which schema?

Reply:
No, I meant to use the :read-config-as-xml operation to get the effective XML.